internal/resource/url.go

```diff
@@ -407,32 +408,73 @@ func (f *Fetcher) fetchFromS3(u url.URL, dest s3target, opts FetchOptions) error
 	}
 	sess := f.AWSSession.Copy()
 
-	// Determine the partition and region this bucket is in
-	regionHint := "us-east-1"
-	if f.S3RegionHint != "" {
-		regionHint = f.S3RegionHint
-	}
-	region, err := s3manager.GetBucketRegion(ctx, sess, u.Host, regionHint)
-	if err != nil {
-		if aerr, ok := err.(awserr.Error); ok && aerr.Code() == "NotFound" {
-			return fmt.Errorf("couldn't determine the region for bucket %q: %v", u.Host, err)
+	// Determine the partition, region, bucket and key based on the URL scheme
+	var bucket, key string
+	switch u.Scheme {
+	case "s3":
+		bucket = u.Host
+		key = u.Path
+
+		// Determine the partition and region this bucket is in
+		regionHint := "us-east-1"
+		if f.S3RegionHint != "" {
+			regionHint = f.S3RegionHint
+		}
+		region, err := s3manager.GetBucketRegion(ctx, sess, bucket, regionHint)
+		if err != nil {
+			if aerr, ok := err.(awserr.Error); ok && aerr.Code() == "NotFound" {
+				return fmt.Errorf("couldn't determine the region for bucket %q: %v", u.Host, err)
+			}
+			return err
 		}
-		return err
-	}
 
-	sess.Config.Region = aws.String(region)
+		sess.Config.Region = aws.String(region)
+	case "arn":
+		fullURL := u.Scheme + ":" + u.Opaque
+		if !arn.IsARN(fullURL) {
+			return configErrors.ErrInvalidS3ARN
+		}
+		s3arn, err := arn.Parse(fullURL)
+		if err != nil {
+			return err
+		}
+		if s3arn.Service != "s3" {
+			return fmt.Errorf("invalid ARN format, must specify an S3 ARN")
+		}
+
+		// Determine if the ARN is for an access point or a bucket.
+		urlSplit := strings.Split(fullURL, "/")
+		if strings.HasPrefix(s3arn.Resource, "accesspoint/") {
+			// When using GetObjectInput with an access point,
+			// you provide the access point ARN in place of the bucket name.
+			// For more information about access point ARNs, see Using access points
+			// https://docs.aws.amazon.com/AmazonS3/latest/userguide/using-access-points.html
+			bucket = strings.Join(urlSplit[:2], "/")
+			key = strings.Join(urlSplit[2:], "/")
+		} else {
+			// Use the full ARN as the bucket name.
+			// If specified, the key is part of the Relative ID which has the format "bucket-name/object-key" according to
+			// https://docs.aws.amazon.com/AmazonS3/latest/userguide/s3-arn-format.html
+			bucket = urlSplit[0]
+			key = strings.Join(urlSplit[1:], "/")
+		}
+
+		sess.Config.Region = aws.String(s3arn.Region)
```

Comment:
S3 ARNs [don't include the Region field](https://docs.aws.amazon.com/AmazonS3/latest/userguide/s3-arn-format.html) so this won't be sufficient for non-`aws` partitions.  We may need to add a hardcoded map from AWS partitions (`s3arn.Partition`) to a representative region for region hinting.  (Or maybe the AWS SDK can do that for us?)

Reply:
AWS partitions are commonly `aws` or `aws-cn` from those docs, and there isn't a clear way to get region from those.

What I've done to address this is to use the current way of using `GetBucketRegion` for the S3 bucket and S3 bucket ARN cases since we have the bucket name available, and used the Region field for the S3 accesspoint case since [the Region field will exist for accesspoints](https://docs.aws.amazon.com/AmazonS3/latest/userguide/using-access-points.html).

Reply:
As is, the non-accesspoint case will fail to select the correct region if we're running in a different partition than the one specified in the ARN.  I still think a lookup table for a representative `regionHint`, indexed by partition, is the right solution here.